README.md
# HashPuppy

Code/ChangeLog
Hash Puppy 0.2
==============
- Added a cancel button.
- Improved layout.
- Start calculating checksum immediately after selecting a file.


Code/main.cpp
/***************************************************************************
 *   Copyright (C) 2009 by AmR EiSa   *
 *   amreisa@gmail.com   *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/
#include <QApplication>
#include "MainWindow.h"

int main( int argc, char *argv[] )
{
    QApplication app( argc, argv );
    MainWindow *window = new MainWindow;
    window->show();
    return app.exec();
}



Code/MainWindow.h
/***************************************************************************
 *   Copyright (C) 2009 by AmR EiSa   *
 *   amreisa@gmail.com   *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/
#ifndef GOTOCELLDIALOG_H
#define GOTOCELLDIALOG_H

#include <QMainWindow>

#include "ui_MainWindow.h"

class MainWindow : public QMainWindow, public Ui::MainWindow
{
    Q_OBJECT
public:
    MainWindow();

protected:
    void dragEnterEvent( QDragEnterEvent *event );
    void dropEvent( QDropEvent *event );

private slots:
    void on_calculateButton_clicked();
    void on_cancelButton_clicked();
    void on_fileEdit_textChanged( const QString &text );
    void on_browseButton_clicked();
    void on_copyButton_clicked();

    void on_actionOpen_triggered();

    void on_actionAbout_triggered();

    void on_comparePushButton_clicked();

private:
    bool cancel_calculation;
};

#endif



Code/main_window.h
/***************************************************************************
 *   Copyright (C) 2009 by Guy Rutenberg   *
 *   guyrutenberg@gmail.com   *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/
#ifndef GOTOCELLDIALOG_H
#define GOTOCELLDIALOG_H

#include <QMainWindow>

#include "ui_main_window.h"

class MainWindow : public QMainWindow, public Ui::MainWindow
{
	Q_OBJECT
public:
	MainWindow();

protected:
	void dragEnterEvent(QDragEnterEvent *event);
	void dropEvent(QDropEvent *event);

private slots:
	void on_calculateButton_clicked();
	void on_cancelButton_clicked();
	void on_fileEdit_textChanged(const QString &text);
	void on_browseButton_clicked();
	void on_copyButton_clicked();

private:
	bool cancel_calculation;
};

#endif



Code/main_window.cpp
/***************************************************************************
 *   Copyright (C) 2009 by Guy Rutenberg   *
 *   guyrutenberg@gmail.com   *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/
#include <QtGui>
#include <QFileDialog>

#include "main_window.h"

MainWindow::MainWindow()
    : QMainWindow()
{
	setupUi(this);
	cancelButton->hide();
	progressBar->hide();

	setFixedHeight(sizeHint().height());

    const QMetaObject & mo = QCryptographicHash::staticMetaObject;
    QStringList hashStrList;
    qDebug() << "Available enums in" << mo.className() << ":" << mo.enumeratorCount();
    int i =0;
    QMetaEnum en = mo.enumerator(i++);
    if ( en.isValid() )
    {
        for ( int num = 0 ; num < en.keyCount() ; num++ ) {
            qDebug() << en.key(num);
            hashStrList << en.key(num);
        }
    }
    else
    {
        qDebug() << "error";
        hashStrList << "error";
    }

    hashSelector->addItems( hashStrList );
}

void MainWindow::on_calculateButton_clicked()
{
	checksumEdit->setText("");

	QFile file(fileEdit->text());
	file.open(QIODevice::ReadOnly);

	const int block_size = (file.size() > 1024*1024) ? 10*1024 : 1024;
	char buffer[10*1024]; // allocate enough space for both cases so msvc will be happy
	int bytes_read;

	int progress_max = file.size()/block_size;
	progress_max = (progress_max > 0) ? progress_max : 1;
	progressBar->setMaximum(progress_max);
	progressBar->reset();

    QCryptographicHash::Algorithm hash_alg = static_cast<QCryptographicHash::Algorithm>(hashSelector->currentIndex());
    QCryptographicHash hash(hash_alg);

	cancel_calculation = false;

	while( (!cancel_calculation && (bytes_read = file.read(buffer, block_size))>0) ) {
		hash.addData(buffer, bytes_read);
		progressBar->setValue(progressBar->value()+1);
		QCoreApplication::processEvents();
	}
	
	progressBar->hide();

	if (!cancel_calculation) {
		checksumEdit->setText(hash.result().toHex());
	}

	progressBar->hide();
	checksumEdit->show();
	
	cancelButton->hide();
	calculateButton->show();
	file.close();
}

void MainWindow::on_cancelButton_clicked()
{
	cancel_calculation = true;
}

void MainWindow::on_fileEdit_textChanged(const QString &text)
{
	QFileInfo f_info(text);
	if(f_info.isFile()) {
		calculateButton->setEnabled(true);
	} else {
		calculateButton->setEnabled(false);
	}
}

void MainWindow::on_browseButton_clicked()
{
	QString filename = QFileDialog::getOpenFileName(this, tr("Load file..."));
	if(!filename.isEmpty()) {
		fileEdit->setText(filename);
	}
	QFileInfo f_info(filename);
	if(f_info.isFile()) {
		calculateButton->click();
	}

}

void MainWindow::on_copyButton_clicked()
{
	QApplication::clipboard()->setText(checksumEdit->text());
}

void MainWindow::dragEnterEvent(QDragEnterEvent *event)
{
	if (event->mimeData()->hasFormat("text/uri-list"))
		event->acceptProposedAction();
}

void MainWindow::dropEvent(QDropEvent *event)
{
	QList<QUrl> urls = event->mimeData()->urls();
	if (urls.isEmpty())
		return;

	QString fileName = urls.first().toLocalFile();
	fileEdit->setText(fileName);
	calculateButton->click();
}


Code/MainWindow.cpp
/***************************************************************************
 *   Copyright (C) 2009 by AmR EiSa   *
 *   amreisa@gmail.com   *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/
#include <QtGui>
#include <QFileDialog>

#include "MainWindow.h"

MainWindow::MainWindow()
    : QMainWindow()
{
    setupUi( this );
    cancelButton->hide();
    progressBar->hide();

    setFixedHeight( sizeHint().height() );

    QStringList hashStrList;
    QMetaEnum en = QCryptographicHash::staticMetaObject.enumerator( 0 );
    if( en.isValid() )
    {
        for( int num = 0 ; num < en.keyCount() ; num++ )
        {
            hashStrList << en.key( num );
        }
    }
    else
    {
        hashStrList << "error";
    }

    hashSelector->addItems( hashStrList );
}

void MainWindow::on_calculateButton_clicked()
{
    checksumEdit->setText( "" );

    QFile file( fileEdit->text() );
    file.open( QIODevice::ReadOnly );

    const int block_size = ( file.size() > 1024*1024 ) ? 10*1024 : 1024;
    char buffer[10*1024]; // allocate enough space for both cases so msvc will be happy
    int bytes_read;

    int progress_max = file.size()/block_size;
    progress_max = ( progress_max > 0 ) ? progress_max : 1;
    progressBar->setMaximum( progress_max );
    progressBar->reset();

    QMetaEnum m = QCryptographicHash::staticMetaObject.enumerator( 0 );

    QCryptographicHash::Algorithm hash_alg = static_cast<QCryptographicHash::Algorithm>( m.keyToValue( hashSelector->currentText().toStdString().c_str() ) );
    QCryptographicHash hash( hash_alg );

    cancel_calculation = false;

    while( ( !cancel_calculation && ( bytes_read = file.read( buffer, block_size ) )>0 ) )
    {
        hash.addData( buffer, bytes_read );
        progressBar->setValue( progressBar->value()+1 );
        QCoreApplication::processEvents();
    }

    progressBar->hide();

    if( !cancel_calculation )
    {
        checksumEdit->setText( hash.result().toHex() );
    }

    progressBar->hide();
    checksumEdit->show();

    cancelButton->hide();
    calculateButton->show();
    file.close();

    comparePushButton->setEnabled( true );
}

void MainWindow::on_cancelButton_clicked()
{
    cancel_calculation = true;
}

void MainWindow::on_fileEdit_textChanged( const QString &text )
{
    QFileInfo f_info( text );
    if( f_info.isFile() )
    {
        calculateButton->setEnabled( true );
    }
    else
    {
        calculateButton->setEnabled( false );
    }
}

void MainWindow::on_browseButton_clicked()
{
    QString filename = QFileDialog::getOpenFileName( this, tr( "Load file..." ) );
    if( !filename.isEmpty() )
    {
        fileEdit->setText( filename );
    }
    QFileInfo f_info( filename );
    if( f_info.isFile() )
    {
        calculateButton->click();
    }

}

void MainWindow::on_copyButton_clicked()
{
    QApplication::clipboard()->setText( checksumEdit->text() );
}

void MainWindow::dragEnterEvent( QDragEnterEvent *event )
{
    if( event->mimeData()->hasFormat( "text/uri-list" ) )
        event->acceptProposedAction();
}

void MainWindow::dropEvent( QDropEvent *event )
{
    QList<QUrl> urls = event->mimeData()->urls();
    if( urls.isEmpty() )
        return;

    QString fileName = urls.first().toLocalFile();
    fileEdit->setText( fileName );
    calculateButton->click();
}

void MainWindow::on_actionOpen_triggered()
{
    emit on_browseButton_clicked();
}

void MainWindow::on_actionAbout_triggered()
{
}

void MainWindow::on_comparePushButton_clicked()
{
    if ( fileEdit->text().isEmpty() )
    {
        // ask for if u want add file
    }
    else
    {
        if ( checksumEdit->text() == textToCompareLineEdit->text() )
            validateLabel->setText( tr( "valid" ) );
    }
//    if ( checksumEdit->text().isEmpty() )
}
